ReportingServicesTools/Functions/Admin/Restore-RsEncryptionKey.ps1

```diff
@@ -117,17 +117,46 @@ function Restore-RSEncryptionKey
         {
             throw "Failed to restore the encryption key! Errors: $($restoreKeyResult.ExtendedErrors)"
         }
-        
+
         try
         {
-            $service = Get-Service -Name $reportServerService -ComputerName $rsWmiObject.PSComputerName -ErrorAction Stop
-            Write-Verbose "Stopping Reporting Services Service... $reportServerService"
-            $service.Stop()
-            $service.WaitForStatus([System.ServiceProcess.ServiceControllerStatus]::Stopped)
-            
-            Write-Verbose "Starting Reporting Services Service... $reportServerService"
-            $service.Start()
-            $service.WaitForStatus([System.ServiceProcess.ServiceControllerStatus]::Running)
+            # Restarting the Reporting Services Windows service requires a different method if a credential is passed because
+            # Get-Service does not have the ability to connect to a remote system with an alternate credential.
+            if ($PSBoundParameters.ContainsKey('Credential'))
+            {
+                $getServiceParams = @{
+                    Class        = 'Win32_Service'
+                    Filter       = "Name = '$reportServerService'"
+                    ComputerName = $rsWmiObject.PSComputerName
+                    Credential   = $Credential
+                }
+                $service = Get-WmiObject @getServiceParams
+
+                Write-Verbose "Stopping Reporting Services Service... $reportServerService"
+                $null = $service.StopService()
+                do {
+                    $service = Get-WmiObject @getServiceParams
+                    Start-Sleep -Seconds 1
+                } until ($service.State -eq 'Stopped')
```

Comment:
I would do a timeout or number of retries , otherwise it had a change to get stuck here in a infinite loop if the service don't stop

Reply:
Good idea, I considered that. However, in the current implementation [WaitForStatus](https://docs.microsoft.com/en-us/dotnet/api/system.serviceprocess.servicecontroller.waitforstatus?view=netframework-4.8#System_ServiceProcess_ServiceController_WaitForStatus_System_ServiceProcess_ServiceControllerStatus_) will wait indefinitely, so I was just mimicking the same behavior.

https://github.com/microsoft/ReportingServicesTools/blob/ffe49b6f1e7dde68503ce94a3ffc87b76d9341c5/ReportingServicesTools/Functions/Admin/Restore-RsEncryptionKey.ps1#L126

What do you think is a reasonable amount of time/retries?

Reply:
I would think a couple of minutes (120 retries) should be good enough



---
In reply to: [282992334](https://github.com/microsoft/ReportingServicesTools/pull/229#discussion_r282992334) [](ancestors = 282992334)